comit/src/expiries.rs

```diff
@@ -0,0 +1,488 @@
+    // This is due to the fact that we must be conservative in our guess of the
+    // counterparties state, hence we may include additional unnecessary
+    // transitions in our calculation. This will only become an issue as we approach
+    // the time cutoff for usefulness i.e., while actors act comfortably within
+    // their 'time to act' window we should not hit this bug.
+    pub fn is_useful<S>(&self, current_state: S) -> bool
+    where
+        S: AsAliceState + AsBobState,
+        C: CurrentTime,
+    {
+        let bob_state = current_state.as_bob_state();
+        let alice_state = current_state.as_alice_state();
+
+        self.alice_has_time_to_complete(alice_state) && self.bob_has_time_to_complete(bob_state)
+    }
+
+    /// True if Alice has time to complete a swap (i.e. transition to done).
+    fn alice_has_time_to_complete(&self, current_state: AliceState) -> bool
+    where
+        C: CurrentTime,
+    {
+        let period = period_for_alice_to_complete(&self.config, current_state);
+        let now = self.connector.beta_current_time();
+
+        // Alice redeems on beta ledger so is concerned about the beta expiry.
+        now + period < self.beta_expiry
+    }
+
+    /// True if Bob has time to complete a swap (i.e transition to done).
+    fn bob_has_time_to_complete(&self, current_state: BobState) -> bool
+    where
+        C: CurrentTime,
+    {
+        let period = period_for_bob_to_complete(&self.config, current_state);
+        let now = self.connector.alpha_current_time();
+
+        // Bob redeems on alpha ledger so is concerned about the alpha expiry.
+        now + period < self.alpha_expiry
+    }
+
+    /// If Alice's next action is not taken within X minutes the expiries will
+    /// become un-useful. Returns X.
+    pub fn alice_should_act_within(&self, current_state: AliceState) -> Duration
+    where
+        C: CurrentTime,
+    {
+        let period = period_for_alice_to_complete(&self.config, current_state);
+        let start_time = self.beta_expiry - period;
+        let now = self.connector.beta_current_time();
+
+        timestamp::duration_between(now, start_time)
+    }
+
+    /// If Bob's next action is not taken within X minutes the expiries will
+    /// become un-useful. Returns X.
+    pub fn bob_should_act_within(&self, current_state: BobState) -> Duration
+    where
+        C: CurrentTime,
+    {
+        let period = period_for_bob_to_complete(&self.config, current_state);
+        let start_time = self.alpha_expiry - period;
+        let now = self.connector.alpha_current_time();
+
+        timestamp::duration_between(now, start_time)
+    }
+
+    /// Returns the recommended next action that Alice should take.
+    pub fn next_action_for_alice(&self, current_state: AliceState) -> AliceAction
+    where
+        C: CurrentTime,
+    {
+        if current_state == AliceState::Done {
+            return AliceAction::NoFurtherAction;
```

Comment:
This variant could be replaced with the use of an `Option`. Not sure if it is better though :)

Reply:
Yeah, I had that at first - I kinda like this solution, not married to it though.